src/py/flwr/cli/log.py

```diff
@@ -0,0 +1,221 @@
+# Copyright 2024 Flower Labs GmbH. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Flower command line interface `log` command."""
+
+import sys
+import time
+from logging import DEBUG, ERROR, INFO
+from pathlib import Path
+from typing import Dict, Optional
+
+import grpc
+import typer
+from typing_extensions import Annotated
+
+from flwr.cli.config_utils import load_and_validate
+from flwr.common.grpc import GRPC_MAX_MESSAGE_LENGTH, create_channel
+from flwr.common.logger import log as logger
+from flwr.proto.exec_pb2 import StreamLogsRequest  # pylint: disable=E0611
+from flwr.proto.exec_pb2_grpc import ExecStub
+
+CONN_REFRESH_PERIOD = 60  # Connection refresh period for log streaming (seconds)
+
+
+def stream_logs(run_id: int, channel: grpc.Channel, duration: int) -> None:
+    """Stream logs from the beginning of a run with connection refresh."""
+    start_time = time.time()
+    stub = ExecStub(channel)
+    req = StreamLogsRequest(run_id=run_id)
+
+    for res in stub.StreamLogs(req):
+        print(res.log_output)
+        if time.time() - start_time > duration:
+            break
+
+
+def print_logs(run_id: int, channel: grpc.Channel, timeout: int) -> None:
+    """Print logs from the beginning of a run."""
+    stub = ExecStub(channel)
+    req = StreamLogsRequest(run_id=run_id)
+
+    while True:
+        try:
+            # Enforce timeout for graceful exit
+            for res in stub.StreamLogs(req, timeout=timeout):
+                print(res.log_output)
+        except grpc.RpcError as e:
+            # pylint: disable=E1101
+            if e.code() == grpc.StatusCode.DEADLINE_EXCEEDED:
+                break
+            if e.code() == grpc.StatusCode.NOT_FOUND:
+                logger(ERROR, "Invalid run_id `%s`, exiting", run_id)
+                break
+            if e.code() == grpc.StatusCode.CANCELLED:
+                break
+
+    channel.close()
```

Comment:
```suggestion
    try:
        while True:
            try:
                # Enforce timeout for graceful exit
                for res in stub.StreamLogs(req, timeout=timeout):
                    print(res.log_output)
            except grpc.RpcError as e:
                # pylint: disable=E1101
                if e.code() == grpc.StatusCode.DEADLINE_EXCEEDED:
                    break
                if e.code() == grpc.StatusCode.NOT_FOUND:
                    logger(ERROR, "Invalid run_id `%s`, exiting", run_id)
                    break
                if e.code() == grpc.StatusCode.CANCELLED:
                    break
    
        channel.close()
    except KeyboardInterrupt:
        logger(DEBUG, "Stream interrupted by user")
    finally:
        channel.close()
        logger(DEBUG, "Channel closed")
```

@Robert-Steiner I guess you mean something like this? @chongshenng I am not 100% sure but I believe this would slightly improve the behaviour.

Reply:
@tanertopal yes exactly. To avoid calling `channel.close()` twice, we should ensure it's called only in the `finally` block.
```python
try:
    while True:
        try:
            # Enforce timeout for graceful exit
            for res in stub.StreamLogs(req, timeout=timeout):
                print(res.log_output)
        except grpc.RpcError as e:
            # pylint: disable=E1101
            if e.code() == grpc.StatusCode.DEADLINE_EXCEEDED:
                break
            if e.code() == grpc.StatusCode.NOT_FOUND:
                logger(ERROR, "Invalid run_id `%s`, exiting", run_id)
                break
            if e.code() == grpc.StatusCode.CANCELLED:
                break
except KeyboardInterrupt:
    logger(DEBUG, "Stream interrupted by user")
finally:
    channel.close()
    logger(DEBUG, "Channel closed")
```

Reply:
Yes, this makes sense, especially in edge cases where there's lots of logs to print. I'll add it.